{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json \n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>clade</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>length</th>\n",
       "      <th>host</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>originating_lab</th>\n",
       "      <th>submitting_lab</th>\n",
       "      <th>authors</th>\n",
       "      <th>url</th>\n",
       "      <th>title</th>\n",
       "      <th>date_submitted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Australia/NSW09/2020</td>\n",
       "      <td>A3</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_413595</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-02-28</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29409</td>\n",
       "      <td>Human</td>\n",
       "      <td>60</td>\n",
       "      <td>Female</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Rockett et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Australia/NSW114/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427724</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-19</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Lam et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Australia/NSW142/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427749</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-23</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Bachmann et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Australia/NSW153/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427653</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-08</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Rockett et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Australia/NSW158/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427735</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-18</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Arnott et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>Wales/PHWC-2B1EE/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_432272</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-07</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Wales</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>29903</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Wales Specialist Virology Centre</td>\n",
       "      <td>Public Health Wales Microbiology Cardiff</td>\n",
       "      <td>Moore et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>Wuhan-Hu-1/2019</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_402125</td>\n",
       "      <td>MN908947</td>\n",
       "      <td>2019-12-26</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29903</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>unknown</td>\n",
       "      <td>National Institute for Communicable Disease Co...</td>\n",
       "      <td>Zhang et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>Wuhan/IPBCAMS-WH-04/2019</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_403929</td>\n",
       "      <td>MT019532</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29890</td>\n",
       "      <td>Human</td>\n",
       "      <td>52</td>\n",
       "      <td>Female</td>\n",
       "      <td>Institute of Pathogen Biology, Chinese Academy...</td>\n",
       "      <td>Institute of Pathogen Biology, Chinese Academy...</td>\n",
       "      <td>Ren et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>Wuhan/WH01/2019</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_406798</td>\n",
       "      <td>LR757998</td>\n",
       "      <td>2019-12-26</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29866</td>\n",
       "      <td>Human</td>\n",
       "      <td>44</td>\n",
       "      <td>Male</td>\n",
       "      <td>General Hospital of Central Theater Command of...</td>\n",
       "      <td>BGI &amp; Institute of Microbiology, Chinese Acade...</td>\n",
       "      <td>Chen et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>Wuhan/WH04/2020</td>\n",
       "      <td>B</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_406801</td>\n",
       "      <td>LR757995</td>\n",
       "      <td>2020-01-05</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29872</td>\n",
       "      <td>Human</td>\n",
       "      <td>39</td>\n",
       "      <td>Male</td>\n",
       "      <td>General Hospital of Central Theater Command of...</td>\n",
       "      <td>BGI &amp; Institute of Microbiology, Chinese Acade...</td>\n",
       "      <td>Chen et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       strain       clade virus  gisaid_epi_isl  \\\n",
       "0        Australia/NSW09/2020          A3  ncov  EPI_ISL_413595   \n",
       "1       Australia/NSW114/2020         A2a  ncov  EPI_ISL_427724   \n",
       "2       Australia/NSW142/2020         A2a  ncov  EPI_ISL_427749   \n",
       "3       Australia/NSW153/2020         A2a  ncov  EPI_ISL_427653   \n",
       "4       Australia/NSW158/2020         A2a  ncov  EPI_ISL_427735   \n",
       "..                        ...         ...   ...             ...   \n",
       "392     Wales/PHWC-2B1EE/2020         A2a  ncov  EPI_ISL_432272   \n",
       "393           Wuhan-Hu-1/2019  unassigned  ncov  EPI_ISL_402125   \n",
       "394  Wuhan/IPBCAMS-WH-04/2019  unassigned  ncov  EPI_ISL_403929   \n",
       "395           Wuhan/WH01/2019  unassigned  ncov  EPI_ISL_406798   \n",
       "396           Wuhan/WH04/2020           B  ncov  EPI_ISL_406801   \n",
       "\n",
       "    genbank_accession        date   region         country         division  \\\n",
       "0                   ?  2020-02-28  Oceania       Australia  New South Wales   \n",
       "1                   ?  2020-03-19  Oceania       Australia  New South Wales   \n",
       "2                   ?  2020-03-23  Oceania       Australia  New South Wales   \n",
       "3                   ?  2020-03-08  Oceania       Australia  New South Wales   \n",
       "4                   ?  2020-03-18  Oceania       Australia  New South Wales   \n",
       "..                ...         ...      ...             ...              ...   \n",
       "392                 ?  2020-04-07   Europe  United Kingdom            Wales   \n",
       "393          MN908947  2019-12-26     Asia           China            Hubei   \n",
       "394          MT019532  2019-12-30     Asia           China            Hubei   \n",
       "395          LR757998  2019-12-26     Asia           China            Hubei   \n",
       "396          LR757995  2020-01-05     Asia           China            Hubei   \n",
       "\n",
       "    location  ... length   host age     sex  \\\n",
       "0     Sydney  ...  29409  Human  60  Female   \n",
       "1     Sydney  ...  29415  Human   ?       ?   \n",
       "2     Sydney  ...  29415  Human   ?       ?   \n",
       "3     Sydney  ...  29415  Human   ?       ?   \n",
       "4     Sydney  ...  29415  Human   ?       ?   \n",
       "..       ...  ...    ...    ...  ..     ...   \n",
       "392      NaN  ...  29903  Human   ?       ?   \n",
       "393    Wuhan  ...  29903  Human   ?       ?   \n",
       "394    Wuhan  ...  29890  Human  52  Female   \n",
       "395    Wuhan  ...  29866  Human  44    Male   \n",
       "396    Wuhan  ...  29872  Human  39    Male   \n",
       "\n",
       "                                       originating_lab  \\\n",
       "0    Centre for Infectious Diseases and Microbiolog...   \n",
       "1    Centre for Infectious Diseases and Microbiolog...   \n",
       "2    Centre for Infectious Diseases and Microbiolog...   \n",
       "3    Centre for Infectious Diseases and Microbiolog...   \n",
       "4    Centre for Infectious Diseases and Microbiolog...   \n",
       "..                                                 ...   \n",
       "392                   Wales Specialist Virology Centre   \n",
       "393                                            unknown   \n",
       "394  Institute of Pathogen Biology, Chinese Academy...   \n",
       "395  General Hospital of Central Theater Command of...   \n",
       "396  General Hospital of Central Theater Command of...   \n",
       "\n",
       "                                        submitting_lab         authors  \\\n",
       "0    NSW Health Pathology - Institute of Clinical P...   Rockett et al   \n",
       "1    NSW Health Pathology - Institute of Clinical P...       Lam et al   \n",
       "2    NSW Health Pathology - Institute of Clinical P...  Bachmann et al   \n",
       "3    NSW Health Pathology - Institute of Clinical P...   Rockett et al   \n",
       "4    NSW Health Pathology - Institute of Clinical P...    Arnott et al   \n",
       "..                                                 ...             ...   \n",
       "392           Public Health Wales Microbiology Cardiff     Moore et al   \n",
       "393  National Institute for Communicable Disease Co...     Zhang et al   \n",
       "394  Institute of Pathogen Biology, Chinese Academy...       Ren et al   \n",
       "395  BGI & Institute of Microbiology, Chinese Acade...      Chen et al   \n",
       "396  BGI & Institute of Microbiology, Chinese Acade...      Chen et al   \n",
       "\n",
       "                        url title date_submitted  \n",
       "0    https://www.gisaid.org     ?     2020-03-08  \n",
       "1    https://www.gisaid.org     ?     2020-04-19  \n",
       "2    https://www.gisaid.org     ?     2020-04-19  \n",
       "3    https://www.gisaid.org     ?     2020-04-19  \n",
       "4    https://www.gisaid.org     ?     2020-04-19  \n",
       "..                      ...   ...            ...  \n",
       "392  https://www.gisaid.org     ?     2020-04-29  \n",
       "393  https://www.gisaid.org     ?     2020-01-12  \n",
       "394  https://www.gisaid.org     ?     2020-01-21  \n",
       "395  https://www.gisaid.org     ?     2020-01-30  \n",
       "396  https://www.gisaid.org     ?     2020-01-30  \n",
       "\n",
       "[397 rows x 24 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = open('results/clades.json',) \n",
    "  \n",
    "# returns JSON object as  \n",
    "# a dictionary \n",
    "data = json.load(f) \n",
    "x = data['nodes']\n",
    "strain = []\n",
    "for i in x:\n",
    "    strain.append(i)  \n",
    "clade = []\n",
    "for y in x.values():\n",
    "    for z in y.keys():\n",
    "        if z == 'clade_membership':\n",
    "            clade.append(y['clade_membership'])\n",
    "list = [strain, clade]\n",
    "df = pd.DataFrame(list, index =['strain', 'clade'])\n",
    "df = df.T\n",
    "drop = []\n",
    "for i in df.index:\n",
    "    if df.loc[i].str.startswith(\"NODE\").values[0]:\n",
    "        drop.append(i)\n",
    "df = df.drop(drop)\n",
    "df_meta = pd.read_csv('data/metadata.tsv', sep='\\t')\n",
    "df = pd.merge(df, df_meta, on=\"strain\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A3\n",
      "0              Australia/NSW09/2020\n",
      "12             Australia/NSW65/2020\n",
      "32           Australia/VIC1233/2020\n",
      "39           Australia/VIC1314/2020\n",
      "74            Australia/VIC645/2020\n",
      "78            Australia/VIC680/2020\n",
      "98            Australia/VIC996/2020\n",
      "161            Georgia/Tb-1523/2020\n",
      "197                 India/1115/2020\n",
      "243       Shandong/IVDC-SD-001/2020\n",
      "251            Shanghai/SH0126/2020\n",
      "375    UnitedArabEmirates/L068/2020\n",
      "Name: strain, dtype: object\n",
      "A2a\n",
      "1      Australia/NSW114/2020\n",
      "2      Australia/NSW142/2020\n",
      "3      Australia/NSW153/2020\n",
      "4      Australia/NSW158/2020\n",
      "16       Australia/NT17/2020\n",
      "               ...          \n",
      "388    Wales/PHWC-275E6/2020\n",
      "389    Wales/PHWC-27C69/2020\n",
      "390    Wales/PHWC-281B4/2020\n",
      "391    Wales/PHWC-29C94/2020\n",
      "392    Wales/PHWC-2B1EE/2020\n",
      "Name: strain, Length: 221, dtype: object\n",
      "B\n",
      "5                Australia/NSW181/2020\n",
      "6                Australia/NSW184/2020\n",
      "7                Australia/NSW193/2020\n",
      "8                Australia/NSW200/2020\n",
      "9                Australia/NSW203/2020\n",
      "10               Australia/NSW207/2020\n",
      "33              Australia/VIC1251/2020\n",
      "35              Australia/VIC1269/2020\n",
      "48                Australia/VIC27/2020\n",
      "54               Australia/VIC330/2020\n",
      "65                Australia/VIC56/2020\n",
      "67               Australia/VIC565/2020\n",
      "100        Beijing/Wuhan_IME-BJ02/2020\n",
      "167    Guangdong/2020XN4459-P0041/2020\n",
      "172            Guangdong/20SF4047/2020\n",
      "179         Guangdong/ZS-S6-P0060/2020\n",
      "182            Guangzhou/GZMU0025/2020\n",
      "192             HongKong/HKU-907a/2020\n",
      "226                 Nanchang/JX14/2020\n",
      "244                Shandong/LY005/2020\n",
      "245               Shanghai/SH0009/2020\n",
      "252                  Singapore/14/2020\n",
      "253                  Singapore/21/2020\n",
      "261             SouthKorea/KCDC05/2020\n",
      "262             SouthKorea/KUMC05/2020\n",
      "264              Spain/Valencia39/2020\n",
      "268                      Taiwan/3/2020\n",
      "272         Thailand/Bangkok-0037/2020\n",
      "396                    Wuhan/WH04/2020\n",
      "Name: strain, dtype: object\n",
      "A1a\n",
      "11         Australia/NSW57/2020\n",
      "13         Australia/NSW75/2020\n",
      "30        Australia/VIC121/2020\n",
      "56        Australia/VIC378/2020\n",
      "64        Australia/VIC556/2020\n",
      "90         Australia/VIC80/2020\n",
      "92         Australia/VIC81/2020\n",
      "111    England/20120002102/2020\n",
      "113    England/20126003706/2020\n",
      "117    England/20130050804/2020\n",
      "120    England/20132082504/2020\n",
      "126    England/20139068804/2020\n",
      "131    England/20144035204/2020\n",
      "137     England/CAMB-74E49/2020\n",
      "144    England/NOTT-10E2A1/2020\n",
      "145    England/NOTT-10E2CF/2020\n",
      "147    England/NOTT-10EA21/2020\n",
      "150     England/SHEF-C0075/2020\n",
      "159           Fuyang/FY002/2020\n",
      "162         Georgia/Tb-468/2020\n",
      "187        Hangzhou/ZJU-01/2020\n",
      "231      Netherlands/NA_65/2020\n",
      "233            Norway/1493/2020\n",
      "241        Scotland/EDB409/2020\n",
      "291            USA/MA_1355/2020\n",
      "381       Wales/PHWC-24F27/2020\n",
      "Name: strain, dtype: object\n",
      "B1\n",
      "14      Australia/NSW96/2020\n",
      "15      Australia/NSW99/2020\n",
      "36     Australia/VIC130/2020\n",
      "43     Australia/VIC160/2020\n",
      "57     Australia/VIC379/2020\n",
      "104      Chongqing/YC01/2020\n",
      "256        Singapore/48/2020\n",
      "275     USA/AZ-TG268903/2020\n",
      "298      USA/MN-MDH-436/2020\n",
      "300         USA/MN_0100/2020\n",
      "313      USA/NY-PV08146/2020\n",
      "336         USA/WA-S137/2020\n",
      "337         USA/WA-S256/2020\n",
      "338         USA/WA-S287/2020\n",
      "339         USA/WA-S391/2020\n",
      "340         USA/WA-S410/2020\n",
      "341         USA/WA-S439/2020\n",
      "342         USA/WA-S505/2020\n",
      "343          USA/WA-S74/2020\n",
      "344          USA/WA-S91/2020\n",
      "348      USA/WA-UW-1805/2020\n",
      "350      USA/WA-UW-1913/2020\n",
      "351      USA/WA-UW-1944/2020\n",
      "359      USA/WA-UW-6164/2020\n",
      "360        USA/WA-UW125/2020\n",
      "361        USA/WA-UW180/2020\n",
      "362        USA/WA-UW229/2020\n",
      "363         USA/WA-UW43/2020\n",
      "364         USA/WA6-UW3/2020\n",
      "368           USA/WI-23/2020\n",
      "Name: strain, dtype: object\n",
      "unassigned\n",
      "49        Australia/VIC276/2020\n",
      "77        Australia/VIC678/2020\n",
      "94        Australia/VIC913/2020\n",
      "102               Brunei/4/2020\n",
      "103               Brunei/5/2020\n",
      "                 ...           \n",
      "384       Wales/PHWC-2571F/2020\n",
      "387       Wales/PHWC-27364/2020\n",
      "393             Wuhan-Hu-1/2019\n",
      "394    Wuhan/IPBCAMS-WH-04/2019\n",
      "395             Wuhan/WH01/2019\n",
      "Name: strain, Length: 72, dtype: object\n",
      "A6\n",
      "108    England/20102112102/2020\n",
      "Name: strain, dtype: object\n",
      "B4\n",
      "165    Guangdong/2020XN4243-P0035/2020\n",
      "281               USA/CT-Yale-088/2020\n",
      "370                     USA/WI-61/2020\n",
      "Name: strain, dtype: object\n",
      "A2\n",
      "180    Guangzhou/GZMU0019/2020\n",
      "181    Guangzhou/GZMU0020/2020\n",
      "247       Shanghai/SH0014/2020\n",
      "Name: strain, dtype: object\n"
     ]
    }
   ],
   "source": [
    "for i in df.clade.unique():\n",
    "    print(i)\n",
    "    print(df[df.clade == i].strain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing input file: None\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_try = df[df.clade == 'A3']\n",
    "df_try.to_csv('A3list.tsv', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexed 399 sequences\n",
      "Selected 12 sequences by ID\n"
     ]
    }
   ],
   "source": [
    "! python scripts/seq_select_by_id.py A3list.tsv 1 results/aligned.fasta FASTA out.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "records = SeqIO.parse(\"out.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'N'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-90-65279081f836>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmotifs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrecords\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mpwm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcounts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnormalize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpseudocounts\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mpssm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpwm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog_odds\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mmatrix_draw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mpssm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'A'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpssm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'C'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpssm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'G'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpssm\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'T'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/nextstrain/lib/python3.6/site-packages/Bio/motifs/__init__.py\u001b[0m in \u001b[0;36mcreate\u001b[0;34m(instances, alphabet)\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0;34m\"\"\"Create a Motif object.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0minstances\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mInstances\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minstances\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mMotif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minstances\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minstances\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/nextstrain/lib/python3.6/site-packages/Bio/motifs/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, alphabet, instances, counts)\u001b[0m\n\u001b[1;32m    271\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minstances\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minstances\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    272\u001b[0m             \u001b[0malphabet\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minstances\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 273\u001b[0;31m             \u001b[0mcounts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minstances\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    274\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcounts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmatrix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFrequencyPositionMatrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malphabet\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcounts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    275\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcounts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/nextstrain/lib/python3.6/site-packages/Bio/motifs/__init__.py\u001b[0m in \u001b[0;36mcount\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    222\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0minstance\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    223\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mposition\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mletter\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minstance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 224\u001b[0;31m                 \u001b[0mcounts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mletter\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mposition\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    225\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mcounts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    226\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'N'"
     ]
    }
   ],
   "source": [
    "m = motifs.create([x.seq for x in records])\n",
    "pwm = m.counts.normalize(pseudocounts=0.5)\n",
    "pssm = pwm.log_odds()\n",
    "matrix_draw = [pssm['A'],pssm['C'],pssm['G'],pssm['T']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
