{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Koleksi spike gene per clade\n",
    "\n",
    "---\n",
    "\n",
    "_oleh: Matin Nuhamunada, M.Sc._\n",
    "\n",
    "Department of Tropical Biology, Universitas Gadjah Mada;   \n",
    "Jl. Teknika Selatan, Sekip Utara, Bulaksumur, Yogyakarta, Indonesia, 55281;   \n",
    "\n",
    "email: [matin_nuhamunada@ugm.ac.id](mailto:matin_nuhamunada@mail.ugm.ac.id)  \n",
    "\n",
    "---\n",
    "#### Notebook Links\n",
    "* 1. [Sub-sampling data genome SARS-CoV-2](01_sub-sampling.ipynb) \n",
    "* 2. [Analisis _genomic epidemiology_ dengan menggunakan Nextstrain](02_analysis.ipynb) \n",
    "* 3. [Koleksi spike gene per clade](03_clade_s_gene_analysis.ipynb) (notebook ini)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deskripsi\n",
    "Pada Notebook ini, akan dilakukan pengambilan sekuen gen spike dari masing-masing clade SARS-CoV-2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Library\n",
    "import json \n",
    "import pandas as pd\n",
    "from os import listdir\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign Clade Information to Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>clade</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>length</th>\n",
       "      <th>host</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>originating_lab</th>\n",
       "      <th>submitting_lab</th>\n",
       "      <th>authors</th>\n",
       "      <th>url</th>\n",
       "      <th>title</th>\n",
       "      <th>date_submitted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Australia/NSW09/2020</td>\n",
       "      <td>A3</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_413595</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-02-28</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29409</td>\n",
       "      <td>Human</td>\n",
       "      <td>60</td>\n",
       "      <td>Female</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Rockett et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Australia/NSW114/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427724</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-19</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Lam et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Australia/NSW142/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427749</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-23</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Bachmann et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Australia/NSW153/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427653</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-08</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Rockett et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Australia/NSW158/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_427735</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-03-18</td>\n",
       "      <td>Oceania</td>\n",
       "      <td>Australia</td>\n",
       "      <td>New South Wales</td>\n",
       "      <td>Sydney</td>\n",
       "      <td>...</td>\n",
       "      <td>29415</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Centre for Infectious Diseases and Microbiolog...</td>\n",
       "      <td>NSW Health Pathology - Institute of Clinical P...</td>\n",
       "      <td>Arnott et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>Wales/PHWC-2B1EE/2020</td>\n",
       "      <td>A2a</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_432272</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-07</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Wales</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>29903</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>Wales Specialist Virology Centre</td>\n",
       "      <td>Public Health Wales Microbiology Cardiff</td>\n",
       "      <td>Moore et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-04-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>Wuhan-Hu-1/2019</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_402125</td>\n",
       "      <td>MN908947</td>\n",
       "      <td>2019-12-26</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29903</td>\n",
       "      <td>Human</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>unknown</td>\n",
       "      <td>National Institute for Communicable Disease Co...</td>\n",
       "      <td>Zhang et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>Wuhan/IPBCAMS-WH-04/2019</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_403929</td>\n",
       "      <td>MT019532</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29890</td>\n",
       "      <td>Human</td>\n",
       "      <td>52</td>\n",
       "      <td>Female</td>\n",
       "      <td>Institute of Pathogen Biology, Chinese Academy...</td>\n",
       "      <td>Institute of Pathogen Biology, Chinese Academy...</td>\n",
       "      <td>Ren et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>Wuhan/WH01/2019</td>\n",
       "      <td>unassigned</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_406798</td>\n",
       "      <td>LR757998</td>\n",
       "      <td>2019-12-26</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29866</td>\n",
       "      <td>Human</td>\n",
       "      <td>44</td>\n",
       "      <td>Male</td>\n",
       "      <td>General Hospital of Central Theater Command of...</td>\n",
       "      <td>BGI &amp; Institute of Microbiology, Chinese Acade...</td>\n",
       "      <td>Chen et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>Wuhan/WH04/2020</td>\n",
       "      <td>B</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_406801</td>\n",
       "      <td>LR757995</td>\n",
       "      <td>2020-01-05</td>\n",
       "      <td>Asia</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>...</td>\n",
       "      <td>29872</td>\n",
       "      <td>Human</td>\n",
       "      <td>39</td>\n",
       "      <td>Male</td>\n",
       "      <td>General Hospital of Central Theater Command of...</td>\n",
       "      <td>BGI &amp; Institute of Microbiology, Chinese Acade...</td>\n",
       "      <td>Chen et al</td>\n",
       "      <td>https://www.gisaid.org</td>\n",
       "      <td>?</td>\n",
       "      <td>2020-01-30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       strain       clade virus  gisaid_epi_isl  \\\n",
       "0        Australia/NSW09/2020          A3  ncov  EPI_ISL_413595   \n",
       "1       Australia/NSW114/2020         A2a  ncov  EPI_ISL_427724   \n",
       "2       Australia/NSW142/2020         A2a  ncov  EPI_ISL_427749   \n",
       "3       Australia/NSW153/2020         A2a  ncov  EPI_ISL_427653   \n",
       "4       Australia/NSW158/2020         A2a  ncov  EPI_ISL_427735   \n",
       "..                        ...         ...   ...             ...   \n",
       "392     Wales/PHWC-2B1EE/2020         A2a  ncov  EPI_ISL_432272   \n",
       "393           Wuhan-Hu-1/2019  unassigned  ncov  EPI_ISL_402125   \n",
       "394  Wuhan/IPBCAMS-WH-04/2019  unassigned  ncov  EPI_ISL_403929   \n",
       "395           Wuhan/WH01/2019  unassigned  ncov  EPI_ISL_406798   \n",
       "396           Wuhan/WH04/2020           B  ncov  EPI_ISL_406801   \n",
       "\n",
       "    genbank_accession        date   region         country         division  \\\n",
       "0                   ?  2020-02-28  Oceania       Australia  New South Wales   \n",
       "1                   ?  2020-03-19  Oceania       Australia  New South Wales   \n",
       "2                   ?  2020-03-23  Oceania       Australia  New South Wales   \n",
       "3                   ?  2020-03-08  Oceania       Australia  New South Wales   \n",
       "4                   ?  2020-03-18  Oceania       Australia  New South Wales   \n",
       "..                ...         ...      ...             ...              ...   \n",
       "392                 ?  2020-04-07   Europe  United Kingdom            Wales   \n",
       "393          MN908947  2019-12-26     Asia           China            Hubei   \n",
       "394          MT019532  2019-12-30     Asia           China            Hubei   \n",
       "395          LR757998  2019-12-26     Asia           China            Hubei   \n",
       "396          LR757995  2020-01-05     Asia           China            Hubei   \n",
       "\n",
       "    location  ... length   host age     sex  \\\n",
       "0     Sydney  ...  29409  Human  60  Female   \n",
       "1     Sydney  ...  29415  Human   ?       ?   \n",
       "2     Sydney  ...  29415  Human   ?       ?   \n",
       "3     Sydney  ...  29415  Human   ?       ?   \n",
       "4     Sydney  ...  29415  Human   ?       ?   \n",
       "..       ...  ...    ...    ...  ..     ...   \n",
       "392      NaN  ...  29903  Human   ?       ?   \n",
       "393    Wuhan  ...  29903  Human   ?       ?   \n",
       "394    Wuhan  ...  29890  Human  52  Female   \n",
       "395    Wuhan  ...  29866  Human  44    Male   \n",
       "396    Wuhan  ...  29872  Human  39    Male   \n",
       "\n",
       "                                       originating_lab  \\\n",
       "0    Centre for Infectious Diseases and Microbiolog...   \n",
       "1    Centre for Infectious Diseases and Microbiolog...   \n",
       "2    Centre for Infectious Diseases and Microbiolog...   \n",
       "3    Centre for Infectious Diseases and Microbiolog...   \n",
       "4    Centre for Infectious Diseases and Microbiolog...   \n",
       "..                                                 ...   \n",
       "392                   Wales Specialist Virology Centre   \n",
       "393                                            unknown   \n",
       "394  Institute of Pathogen Biology, Chinese Academy...   \n",
       "395  General Hospital of Central Theater Command of...   \n",
       "396  General Hospital of Central Theater Command of...   \n",
       "\n",
       "                                        submitting_lab         authors  \\\n",
       "0    NSW Health Pathology - Institute of Clinical P...   Rockett et al   \n",
       "1    NSW Health Pathology - Institute of Clinical P...       Lam et al   \n",
       "2    NSW Health Pathology - Institute of Clinical P...  Bachmann et al   \n",
       "3    NSW Health Pathology - Institute of Clinical P...   Rockett et al   \n",
       "4    NSW Health Pathology - Institute of Clinical P...    Arnott et al   \n",
       "..                                                 ...             ...   \n",
       "392           Public Health Wales Microbiology Cardiff     Moore et al   \n",
       "393  National Institute for Communicable Disease Co...     Zhang et al   \n",
       "394  Institute of Pathogen Biology, Chinese Academy...       Ren et al   \n",
       "395  BGI & Institute of Microbiology, Chinese Acade...      Chen et al   \n",
       "396  BGI & Institute of Microbiology, Chinese Acade...      Chen et al   \n",
       "\n",
       "                        url title date_submitted  \n",
       "0    https://www.gisaid.org     ?     2020-03-08  \n",
       "1    https://www.gisaid.org     ?     2020-04-19  \n",
       "2    https://www.gisaid.org     ?     2020-04-19  \n",
       "3    https://www.gisaid.org     ?     2020-04-19  \n",
       "4    https://www.gisaid.org     ?     2020-04-19  \n",
       "..                      ...   ...            ...  \n",
       "392  https://www.gisaid.org     ?     2020-04-29  \n",
       "393  https://www.gisaid.org     ?     2020-01-12  \n",
       "394  https://www.gisaid.org     ?     2020-01-21  \n",
       "395  https://www.gisaid.org     ?     2020-01-30  \n",
       "396  https://www.gisaid.org     ?     2020-01-30  \n",
       "\n",
       "[397 rows x 24 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = open('results/clades.json',) \n",
    "  \n",
    "data = json.load(f) \n",
    "x = data['nodes']\n",
    "strain = []\n",
    "for i in x:\n",
    "    strain.append(i)  \n",
    "clade = []\n",
    "for y in x.values():\n",
    "    for z in y.keys():\n",
    "        if z == 'clade_membership':\n",
    "            clade.append(y['clade_membership'])\n",
    "list = [strain, clade]\n",
    "df = pd.DataFrame(list, index =['strain', 'clade'])\n",
    "df = df.T\n",
    "drop = []\n",
    "for i in df.index:\n",
    "    if df.loc[i].str.startswith(\"NODE\").values[0]:\n",
    "        drop.append(i)\n",
    "df = df.drop(drop)\n",
    "df_meta = pd.read_csv('data/metadata.tsv', sep='\\t')\n",
    "df = pd.merge(df, df_meta, on=\"strain\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check number of samples per clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clade A3\n",
      "Australia               7\n",
      "China                   2\n",
      "United Arab Emirates    1\n",
      "Georgia                 1\n",
      "India                   1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade A2a\n",
      "USA               68\n",
      "Australia         65\n",
      "United Kingdom    44\n",
      "Netherlands        5\n",
      "India              4\n",
      "France             4\n",
      "Jordan             3\n",
      "Singapore          3\n",
      "Luxembourg         3\n",
      "Hungary            2\n",
      "Georgia            2\n",
      "Thailand           2\n",
      "Sweden             2\n",
      "Denmark            2\n",
      "Russia             1\n",
      "Turkey             1\n",
      "Austria            1\n",
      "Portugal           1\n",
      "Spain              1\n",
      "Czech Republic     1\n",
      "Finland            1\n",
      "Belgium            1\n",
      "Indonesia          1\n",
      "Taiwan             1\n",
      "Iceland            1\n",
      "Germany            1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade B\n",
      "Australia      12\n",
      "China           9\n",
      "Singapore       2\n",
      "South Korea     2\n",
      "Spain           1\n",
      "Hong Kong       1\n",
      "Thailand        1\n",
      "Taiwan          1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade A1a\n",
      "United Kingdom    13\n",
      "Australia          7\n",
      "China              2\n",
      "Netherlands        1\n",
      "USA                1\n",
      "Norway             1\n",
      "Georgia            1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade B1\n",
      "USA          23\n",
      "Australia     5\n",
      "Singapore     1\n",
      "China         1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade unassigned\n",
      "China             26\n",
      "United Kingdom     9\n",
      "Indonesia          8\n",
      "India              7\n",
      "USA                6\n",
      "Hong Kong          4\n",
      "Singapore          3\n",
      "Australia          3\n",
      "Philippines        2\n",
      "Brunei             2\n",
      "Thailand           1\n",
      "Taiwan             1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade A6\n",
      "United Kingdom    1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade B4\n",
      "USA      2\n",
      "China    1\n",
      "Name: country, dtype: int64\n",
      " \n",
      "clade A2\n",
      "China    3\n",
      "Name: country, dtype: int64\n",
      " \n"
     ]
    }
   ],
   "source": [
    "for i in df.clade.unique():\n",
    "    print(\"clade \"+i)\n",
    "    print(df[df.clade == i].country.value_counts())\n",
    "    print(\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Fasta per Clade\n",
    "Tribute to Peter Cock, The James Hutton Institute for the script: https://github.com/peterjc/pico_galaxy/tree/master/tools/seq_select_by_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexed 399 sequences\n",
      "Selected 12 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 221 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 29 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 26 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 30 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 72 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 1 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 3 sequences by ID\n",
      "Indexed 399 sequences\n",
      "Selected 3 sequences by ID\n"
     ]
    }
   ],
   "source": [
    "for i in df.clade.unique():\n",
    "    df_try = df[df.clade == i]\n",
    "    df_try.to_csv('results/'+i+'_clade_list.tsv', header=False, index=False, sep='\\t')\n",
    "    ! python scripts/seq_select_by_id.py results/{i}_clade_list.tsv 1 results/aligned.fasta FASTA results/clade/{i}.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract Fasta from Indonesian sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexed 399 sequences\n",
      "Selected 9 sequences by ID\n"
     ]
    }
   ],
   "source": [
    "df_try = df[df.country == \"Indonesia\"]\n",
    "df_try.to_csv('indonesia.tsv', header=False, index=False, sep='\\t')\n",
    "! python scripts/seq_select_by_id.py indonesia.tsv 1 results/aligned.fasta FASTA results/clade/indonesia.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select only Spike Gene from genome sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_spike_clade(file, name):\n",
    "    seq_reference = SeqIO.read(\"config/reference.gb\", \"genbank\")\n",
    "    for feature in seq_reference.features:\n",
    "        for x in feature.qualifiers:\n",
    "            if x == 'gene':\n",
    "                if feature.qualifiers['gene'][0] == \"S\":\n",
    "                    start = feature.location.start.position\n",
    "                    end = feature.location.end.position\n",
    "    gene_records_full = []\n",
    "    gene_id_full = []\n",
    "    gene_records = []\n",
    "    gene_id = []\n",
    "\n",
    "    for records in SeqIO.parse(file, \"fasta\"):\n",
    "        sequence = str(records.seq[start:end]).upper()\n",
    "        gene_records_full.append(records.seq[start:end])\n",
    "        gene_id_full.append(records.id)\n",
    "        amb = float(sequence.count(\"N\")) + float(sequence.count(\"Y\"))\n",
    "        \n",
    "        if amb > 0 :\n",
    "            print(str(records.id)+\" \"+str(amb))\n",
    "        else:            \n",
    "            gene_records.append(records.seq[start:end])\n",
    "            gene_id.append(records.id)\n",
    "    \n",
    "    records = []\n",
    "    for i in range(len(gene_id_full)):\n",
    "        records.append(SeqRecord(gene_records_full[i], gene_id_full[i]))\n",
    "    SeqIO.write(records, \"results/clade_spike/\"+name+\"_spike_.faa\", \"fasta\")\n",
    "    \n",
    "    return gene_records_full, gene_id_full, gene_records, gene_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Print sequence with low quality (ambiguous nucleotide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A1a\n",
      "Australia/VIC121/2020 97.0\n",
      "Australia/VIC556/2020 149.0\n",
      "England/20126003706/2020 184.0\n",
      "England/20130050804/2020 84.0\n",
      "England/NOTT-10E2A1/2020 249.0\n",
      "England/NOTT-10E2CF/2020 87.0\n",
      "\n",
      "A2\n",
      "\n",
      "A2a\n",
      "Australia/NT17/2020 13.0\n",
      "Australia/VIC1020/2020 458.0\n",
      "Australia/VIC1033/2020 486.0\n",
      "Australia/VIC1049/2020 447.0\n",
      "Australia/VIC1054/2020 388.0\n",
      "Australia/VIC107/2020 7.0\n",
      "Australia/VIC1105/2020 588.0\n",
      "Australia/VIC1117/2020 893.0\n",
      "Australia/VIC1174/2020 150.0\n",
      "Australia/VIC1209/2020 435.0\n",
      "Australia/VIC1224/2020 186.0\n",
      "Australia/VIC1256/2020 185.0\n",
      "Australia/VIC1313/2020 2.0\n",
      "Australia/VIC1337/2020 5.0\n",
      "Australia/VIC1355/2020 199.0\n",
      "Australia/VIC197/2020 1.0\n",
      "Australia/VIC420/2020 90.0\n",
      "Australia/VIC421/2020 184.0\n",
      "Australia/VIC438/2020 185.0\n",
      "Australia/VIC490/2020 182.0\n",
      "Australia/VIC503/2020 30.0\n",
      "Australia/VIC536/2020 2.0\n",
      "Australia/VIC582/2020 85.0\n",
      "Australia/VIC583/2020 163.0\n",
      "Australia/VIC616/2020 149.0\n",
      "Australia/VIC619/2020 185.0\n",
      "Australia/VIC624/2020 199.0\n",
      "Australia/VIC625/2020 122.0\n",
      "Australia/VIC664/2020 5.0\n",
      "Australia/VIC673/2020 204.0\n",
      "Australia/VIC720/2020 194.0\n",
      "Australia/VIC725/2020 185.0\n",
      "Australia/VIC748/2020 135.0\n",
      "Australia/VIC755/2020 210.0\n",
      "Australia/VIC756/2020 652.0\n",
      "Australia/VIC762/2020 59.0\n",
      "Australia/VIC774/2020 186.0\n",
      "Australia/VIC794/2020 170.0\n",
      "Australia/VIC798/2020 114.0\n",
      "Australia/VIC799/2020 125.0\n",
      "Australia/VIC800/2020 58.0\n",
      "Australia/VIC971/2020 188.0\n",
      "Australia/VIC973/2020 624.0\n",
      "Belgium/UGent-32/2020 218.0\n",
      "CzechRepublic/2282/2020 30.0\n",
      "England/20106005303/2020 184.0\n",
      "England/20109050506/2020 1.0\n",
      "England/20129163604/2020 1.0\n",
      "England/20132074102/2020 184.0\n",
      "England/201360276/2020 75.0\n",
      "England/20136087004/2020 9.0\n",
      "England/20138007704/2020 184.0\n",
      "England/20142026502/2020 91.0\n",
      "England/20144000404/2020 1.0\n",
      "England/CAMB-73800/2020 1.0\n",
      "England/CAMB-73A37/2020 3.0\n",
      "England/CAMB-74517/2020 225.0\n",
      "England/CAMB-78197/2020 120.0\n",
      "England/CAMB-7C209/2020 594.0\n",
      "England/NOTT-10E890/2020 1.0\n",
      "England/SHEF-C5ABE/2020 485.0\n",
      "England/SHEF-D30F7/2020 267.0\n",
      "Jordan/SR-055/2020 307.0\n",
      "Scotland/EDB020/2020 1.0\n",
      "USA/FL-BPHL-016/2020 312.0\n",
      "USA/FL-BPHL-028/2020 240.0\n",
      "USA/MI-MDHHS-SC20185/2020 8.0\n",
      "USA/MN-MDH-117/2020 1.0\n",
      "USA/MN132-MDH132/2020 1.0\n",
      "USA/NY-NYUMC93/2020 126.0\n",
      "USA/NY-PV09080/2020 427.0\n",
      "USA/SEARCH-0058-NBG/2020 22.0\n",
      "USA/TX-HMH058/2020 220.0\n",
      "USA/WA-UW-1574/2020 9.0\n",
      "USA/WI-UW-273/2020 1.0\n",
      "Wales/PHW21/2020 267.0\n",
      "\n",
      "A3\n",
      "Australia/VIC1233/2020 324.0\n",
      "Australia/VIC1314/2020 1.0\n",
      "Australia/VIC996/2020 120.0\n",
      "\n",
      "A6\n",
      "\n",
      "B\n",
      "Australia/VIC1251/2020 238.0\n",
      "Australia/VIC1269/2020 646.0\n",
      "Australia/VIC330/2020 7.0\n",
      "Guangdong/ZS-S6-P0060/2020 267.0\n",
      "Thailand/Bangkok-0037/2020 29.0\n",
      "\n",
      "B1\n",
      "Australia/VIC379/2020 93.0\n",
      "Singapore/48/2020 21.0\n",
      "\n",
      "B4\n",
      "USA/CT-Yale-088/2020 3.0\n",
      "\n",
      "indonesia\n",
      "\n",
      "unassigned\n",
      "Australia/VIC678/2020 13.0\n",
      "Australia/VIC913/2020 185.0\n",
      "Brunei/4/2020 14.0\n",
      "England/CAMB-75419/2020 218.0\n",
      "England/NOTT-10F0D3/2020 2.0\n",
      "Guangdong/2020XN4433-P0040/2020 13.0\n",
      "Guangdong/20SF156/2020 642.0\n",
      "Guangdong/20SF822/2020 38.0\n",
      "Guangdong/FS-S50-P0053/2020 267.0\n",
      "Guangdong/GDFS2020052-P0025/2020 267.0\n",
      "Guangdong/GDSZ202008-P0020/2020 15.0\n",
      "HongKong/HKPU61-0202/2020 267.0\n",
      "HongKong/HKPU91-2002/2020 1.0\n",
      "India/NCDC-02331/2020 4.0\n",
      "India/NCDC-02336/2020 3.0\n",
      "India/NCDC-2535/2020 1.0\n",
      "India/NCDC-3190/2020 1.0\n",
      "India/NCDC-3394/2020 1.0\n",
      "India/NCDC-3934/2020 3.0\n",
      "India/NCDC-4189/2020 1.0\n",
      "Philippines/RITM-06/2020 218.0\n",
      "Shanghai/SH0104/2020 109.0\n",
      "USA/CA-CZB-1052/2020 44.0\n",
      "USA/CA-SCCPHD-UC138/2020 408.0\n",
      "Wales/PHWC-23FBF/2020 1.0\n",
      "Wales/PHWC-253C0/2020 3.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fasta_clade = listdir(\"results/clade/\")\n",
    "for i in range(len(fasta_clade)):\n",
    "    if '.fasta' in fasta_clade[i]:\n",
    "        file = \"results/clade/\"+fasta_clade[i]\n",
    "        name = fasta_clade[i].replace('.fasta', '')\n",
    "        print(name)\n",
    "        get_spike_clade(file, name)\n",
    "        print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_bit_matrix(m, start, stop, title):\n",
    "    pwm = m.counts.normalize(pseudocounts=0.5)\n",
    "    pssm = pwm.log_odds()\n",
    "    matrix_draw = [pssm['A'],pssm['C'],pssm['G'],pssm['T']]\n",
    "    \n",
    "    # set the size of the figure\n",
    "    plt.figure(figsize=[20,2])\n",
    "\n",
    "    # show the array flipped (transposed) and with no colour interpolation smoothing\n",
    "    plt.imshow(matrix_draw,interpolation='nearest')\n",
    "\n",
    "    # set the ticks\n",
    "    #startx, endx = plt.get_xlim()\n",
    "    plt.xticks(range(start,stop, 5))\n",
    "    plt.xlim([start, stop])\n",
    "    plt.yticks(range(4),['A','C','G','T'])\n",
    "\n",
    "    # set the colorbar\n",
    "    plt.clim()\n",
    "    plt.colorbar()\n",
    "\n",
    "    # title\n",
    "    plt.title('base bit score matrix of '+title+' position '+str(start)+\":\"+str(stop),fontsize=10)\n",
    "\n",
    "    # show the figure\n",
    "    return plt.show()\n",
    "\n",
    "def bit_matrix(records, title):\n",
    "    print(\"Number of samples: \"+str(len(records)))\n",
    "    m = motifs.create(gene_records)\n",
    "    number = int(len(m)/100)\n",
    "    start = []\n",
    "    stop = []\n",
    "    for i in range(number):\n",
    "        n = (i*100)\n",
    "        s = ((i+1)*100-1)\n",
    "        start.append(n)\n",
    "        stop.append(s)\n",
    "    for i in range(len(start)):\n",
    "        draw_bit_matrix(m, start[i], stop[i], title)\n",
    "    return m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_records' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-379cff6d9cc1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbit_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_records\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"spike gene\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'gene_records' is not defined"
     ]
    }
   ],
   "source": [
    "m = bit_matrix(gene_records, \"spike gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = motifs.create(gene_records)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
