{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Library\n",
    "from pandas import DataFrame\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez \n",
    "from Bio import SeqIO\n",
    "import pycurl\n",
    "import matplotlib.pyplot as plt\n",
    "import os, sys\n",
    "from tqdm.notebook import trange, tqdm\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    df = pd.read_csv(\"output/ncov2019_ncbi_spike.csv\", index_col=0)\n",
    "except:\n",
    "    df = pd.read_csv(\"dataset/ncov2019_ncbi08052020.csv\")\n",
    "    df[\"s-gene_location\"] = \"\"\n",
    "    df[\"gene_length\"] = \"\"\n",
    "    df[\"protein_id\"] = \"\"\n",
    "    df[\"protein_length\"] = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Length</th>\n",
       "      <th>Geo_Location</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>s-gene_location</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512</td>\n",
       "      <td>2020-01-13T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29903</td>\n",
       "      <td>China</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-12</td>\n",
       "      <td>[21562:25384](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>YP_009724390.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT444148</td>\n",
       "      <td>2020-05-08T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29890</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>oronasopharynx</td>\n",
       "      <td>2020-03-11</td>\n",
       "      <td>[21549:25371](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>QJQ82624.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT444515</td>\n",
       "      <td>2020-05-08T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29804</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-03-16</td>\n",
       "      <td>[21530:25352](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>QJQ82636.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT444516</td>\n",
       "      <td>2020-05-08T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29786</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-03-16</td>\n",
       "      <td>[21512:25334](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>QJQ82648.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT444517</td>\n",
       "      <td>2020-05-08T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29797</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-03-16</td>\n",
       "      <td>[21523:25345](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>QJQ82660.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td>MN975268</td>\n",
       "      <td>2020-01-24T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>107</td>\n",
       "      <td>China</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>lung, oronasopharynx</td>\n",
       "      <td>2020-01</td>\n",
       "      <td>[&lt;0:&gt;107](+)</td>\n",
       "      <td>107.0</td>\n",
       "      <td>QHN73824.1</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2430</th>\n",
       "      <td>MN985325</td>\n",
       "      <td>2020-01-24T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29882</td>\n",
       "      <td>USA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>oronasopharynx</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>[21562:25384](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>QHO60594.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2431</th>\n",
       "      <td>MN970003</td>\n",
       "      <td>2020-01-23T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>290</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>lung, oronasopharynx</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2432</th>\n",
       "      <td>MN970004</td>\n",
       "      <td>2020-01-23T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>290</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>lung, oronasopharynx</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>MN908947</td>\n",
       "      <td>2020-01-12T00:00:00Z</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>29903</td>\n",
       "      <td>China</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-12</td>\n",
       "      <td>[21562:25384](+)</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>QHD43416.1</td>\n",
       "      <td>1273.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2434 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Accession          Release_Date  \\\n",
       "0     NC_045512  2020-01-13T00:00:00Z   \n",
       "1      MT444148  2020-05-08T00:00:00Z   \n",
       "2      MT444515  2020-05-08T00:00:00Z   \n",
       "3      MT444516  2020-05-08T00:00:00Z   \n",
       "4      MT444517  2020-05-08T00:00:00Z   \n",
       "...         ...                   ...   \n",
       "2429   MN975268  2020-01-24T00:00:00Z   \n",
       "2430   MN985325  2020-01-24T00:00:00Z   \n",
       "2431   MN970003  2020-01-23T00:00:00Z   \n",
       "2432   MN970004  2020-01-23T00:00:00Z   \n",
       "2433   MN908947  2020-01-12T00:00:00Z   \n",
       "\n",
       "                                                Species  Length Geo_Location  \\\n",
       "0     Severe acute respiratory syndrome-related coro...   29903        China   \n",
       "1     Severe acute respiratory syndrome-related coro...   29890      USA: CA   \n",
       "2     Severe acute respiratory syndrome-related coro...   29804          USA   \n",
       "3     Severe acute respiratory syndrome-related coro...   29786          USA   \n",
       "4     Severe acute respiratory syndrome-related coro...   29797          USA   \n",
       "...                                                 ...     ...          ...   \n",
       "2429  Severe acute respiratory syndrome-related coro...     107        China   \n",
       "2430  Severe acute respiratory syndrome-related coro...   29882          USA   \n",
       "2431  Severe acute respiratory syndrome-related coro...     290     Thailand   \n",
       "2432  Severe acute respiratory syndrome-related coro...     290     Thailand   \n",
       "2433  Severe acute respiratory syndrome-related coro...   29903        China   \n",
       "\n",
       "              Host      Isolation_Source Collection_Date   s-gene_location  \\\n",
       "0     Homo sapiens                   NaN         2019-12  [21562:25384](+)   \n",
       "1     Homo sapiens        oronasopharynx      2020-03-11  [21549:25371](+)   \n",
       "2     Homo sapiens                   NaN      2020-03-16  [21530:25352](+)   \n",
       "3     Homo sapiens                   NaN      2020-03-16  [21512:25334](+)   \n",
       "4     Homo sapiens                   NaN      2020-03-16  [21523:25345](+)   \n",
       "...            ...                   ...             ...               ...   \n",
       "2429  Homo sapiens  lung, oronasopharynx         2020-01      [<0:>107](+)   \n",
       "2430  Homo sapiens        oronasopharynx      2020-01-19  [21562:25384](+)   \n",
       "2431  Homo sapiens  lung, oronasopharynx      2020-01-08               NaN   \n",
       "2432  Homo sapiens  lung, oronasopharynx      2020-01-13               NaN   \n",
       "2433  Homo sapiens                   NaN         2019-12  [21562:25384](+)   \n",
       "\n",
       "      gene_length      protein_id  protein_length  \n",
       "0          3822.0  YP_009724390.1          1273.0  \n",
       "1          3822.0      QJQ82624.1          1273.0  \n",
       "2          3822.0      QJQ82636.1          1273.0  \n",
       "3          3822.0      QJQ82648.1          1273.0  \n",
       "4          3822.0      QJQ82660.1          1273.0  \n",
       "...           ...             ...             ...  \n",
       "2429        107.0      QHN73824.1            35.0  \n",
       "2430       3822.0      QHO60594.1          1273.0  \n",
       "2431          NaN             NaN             NaN  \n",
       "2432          NaN             NaN             NaN  \n",
       "2433       3822.0      QHD43416.1          1273.0  \n",
       "\n",
       "[2434 rows x 12 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c31fa57c932e4633911008f5f49c0fd4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=2434.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "error\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Email for NCBI\n",
    "Entrez.email = \"matin_nuhamunada@ugm.ac.id\"\n",
    "\n",
    "for i in tqdm(df.index):\n",
    "    if df.loc[i, \"s-gene_location\"] == '':\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=df.Accession[i], rettype=\"gb\", retmode=\"text\")\n",
    "        seq_record = SeqIO.read(handle, \"genbank\")\n",
    "        handle.close()\n",
    "        #print(seq_record.id)\n",
    "        for feature in seq_record.features:\n",
    "            try:\n",
    "                if feature.type == 'CDS':\n",
    "                    if feature.qualifiers['gene'][0] == 'S':\n",
    "                        start = feature.location.start.position\n",
    "                        end = feature.location.end.position\n",
    "                        protein_id = str(feature.qualifiers['protein_id'][0])\n",
    "                \n",
    "                        df.loc[i, \"s-gene_location\"] = str(feature.location)\n",
    "                        df.loc[i, \"gene_length\"] = len(seq_record.seq[start:end])\n",
    "                        df.loc[i, \"protein_id\"] = protein_id\n",
    "                        df.loc[i, \"protein_length\"] = len(feature.qualifiers['translation'][0])\n",
    "                \n",
    "                    #write spike gene fasta\n",
    "                        record = SeqRecord(Seq(\"\"))\n",
    "                        record.seq = seq_record.seq[start:end]\n",
    "                        record.id = seq_record.id\n",
    "                        with open(\"output/gene/s_gene_\"+str(i)+\"_\"+record.id+\".fasta\", \"w\") as output_handle:\n",
    "                            SeqIO.write(record, output_handle, \"fasta\")\n",
    "                \n",
    "                    #write spike protein fasta\n",
    "                        protein = feature.qualifiers['translation'][0]\n",
    "                        protein_record = SeqRecord(Seq(protein, IUPAC.protein), id= protein_id, description=feature.qualifiers['product'][0])\n",
    "                        with open(\"output/protein/spike_\"+str(i)+\"_\"+protein_id+\".fasta\", \"w\") as output_handle:\n",
    "                            SeqIO.write(protein_record, output_handle, \"fasta\")\n",
    "            except:\n",
    "                print('error')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('output/ncov2019_ncbi_spike.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
